### ISSUE TITLE
Expose input-element-value to facilitate AJAX requests

### ISSUE THREAD

**Issue by mercmobily** [vaadin/vaadin-combo-box#284]
At the moment, I would like to do something like this:

```
<iron-ajax auto={{searchName}} url="/stores/polymer?nameStarts=[[searchName]]" last-response={{ajaxResponse}}></iron-ajax>
<vaadin-combo-box id="VCC" input-element-value="{{searchName}}" label="Language" class="languages" items="[[ajaxResponse]]"></vaadin-combo-box>
```

Basically, the AJAX request is initiated when something is typed. iron-ajax makes the request to the server, and `{{ajaxResponse}}` is bound to the last response. That bound variable is fed to the wonderful vaadin-combo-box, and voila' -- only the "legal" options are available. Nirvana of auto-completion is reached.

But! vaadin-combo-box doesn't expose input-element-value. The change is trivial: just add:

```
  inputElementValue: {
    type: String,
    notify: true
  },
```

And then:

```
 /**
 *  Filtering and items handling
 */

_inputValueChanged: function(e) {

  // Expose inputElementValue
  this.inputElementValue = this._inputElementValue;
```

To `vaadin-combo-box-behavior.html`. These two changes basically mean that the code above works. I tried a million possible ways to avoid changing vaadin-checkbox, but no -- if the change isn't exposed, it's just not going to work!

For the sake of making AJAX autocompletion work, would you accept a pull request with the changes above?

Another option would be to rename `_inputElementValue` into `inputElementValue`. However, this would be less than ideal. Really, in terms of facilitating AJAX requests, we should only change `inputElementValue` if there has been an `input` event on the input field. So, in terms of behaviour in terms of making an AJAX request, the way I proposed is _perfect_.

Maybe inputElementValue isn't _actually_ the best name for it. It should be something like `keyedInValue` or... something that clarifies that it's the result of user typing (hence the auto-completion query).

The only way I have to go about it is to write a decorator widget that mirrors _every_ possible input of `vaadin-combo-box.html`, and uses an extra behaviour that overwrites `_inputValueChanged`. But that really is way less than ideal...! Alternatively, I can clone `vaadin-combo-box.html` so that it redefines `_inputValueChanged` but even then...

It would be awesome if `vaadin-combo-box` was usable with AJAX-loaded data out of the box. Especially since the modifications would be so minimal...

Thanks for listening!

**Comment by Saulis** [comment]
Hi there!

Have you considered using `vaadin-combo-box-light` to expose the whole input element? 

Please see https://vaadin.com/docs/-/part/elements/vaadin-combo-box/vaadin-combo-box-customization.html for details.

**Comment by mercmobily** [comment]
I think that's a very workable idea. But.. I really like the vaadin-combo-box! Don't you think that 3 lines of code would be worth make the element totally workable with Ajax calls? It just makes so much sense... I would be happy to send a PR with documentation and tests...

**Comment by Saulis** [comment]
We try to generally be careful when exposing new API since removing it afterwards might be difficult to do. What we're looking for in new API is that it solves some relevant use case that can't be done with the current API – or is currently really complex to do.

I think the use case here is relevant. However, I'm not sure if you should be using the input value. You see, the input value gets changed also when a user navigates the suggestions using arrow keys.

Instead, I believe you need to listen to `input` event on the input element to trigger the fetching of items. Currently, that can be done either by manually adding a listener to `combobox.$.input` or exposing the input element by using `vaadin-combo-box-light` and `on-input` handler.

Now, although there are in my view at least two ways to achieve the goal, there isn't really an easy way to make it happen purely declaratively. So there is room for improvement. 

As it happens, there is a private property `_filter` which I think would fit the purpose here. In ideal case, it just needs `notify:true` to be added. Can't promise I have time to test it out before next week, but if you're interested and able to help out, we would really appreciate it.

**Comment by mercmobily** [comment]
I did a lot more practical, real-life testing on this. While my original idea seems good "in theory", using this widget for auto-completion (which is what I am basically doing) is like using an adjustable spanner to undo a bolt: while it's _possible_, it's much less than ideal and it doesn't give really good results.

I looked at the code, and I agree: the `_filter` property does seem to be the way to go. I would be happy to help you test it, no worries at all. However, things like:

* Adding selected text to the rest of the input box like the best auto-completions do
* Keyboard navigation, which requires interaction between the input box and the select

Are probably out of the scope of this widget.

I am developing my own. I will send you a link, so you will see what I mean. But, totally happy to help out to make vaadin-combo-box more AJAX-friendly!

**Comment by jouni** [comment]
Hi! The plan is definitely so that we can support the autocomplete use case as well as the refular combo box case. It’s been in the backlog since the very befinning, but we haven’t had the time to het it done yet.

Whether it is a separate element from vaadin-combo-box is yet to be determined, but it will definitely share use vaadin-combo-box as its foundation. And both of the features you mentioned in the bullet points are relevant. So please, we’re more than glad to get your help on this, even just as a tester if not a direct code contributor.

**Comment by mercmobily** [comment]
Hi Jouni,

This is my attempt to get autocomplete going:

https://github.com/mercmobily/hot-autocomplete

Notes:

* I coded the keyboard navigation as best as I could. Focus is tricky. I found it really, really difficult to make sure that when the input field loses the focus, the suggestion bar should disappear. I think there must be better ways of doing it

* I created a surrounding widget, that will work on the first widget. Now that I think about it, a much better Polymer-ish way of doing it would be to assign a specific ID to the widget, and use that ID to distribute the right content. Or maybe not?

* The server really needs to cooperate

I hope this can be half useful... but I am TOTALLY here, willing to help out to making vaadin-combo-box work.

BTW, I am totally happy for you to take my code (and I mean *take*: your copyright, etc.) if you decide to take bits & pieces, or if you decide to make a "proper" autocomplete widget!

**Comment by jouni** [comment]
Thanks for the effort, I appreciate it! Can’t promise exactly when the team will have time to take a closer look at the autocomplete feature for vaadin-combo-box, but we’ll let you know!

**Comment by panuhorsmalahti** [comment]
I'm also looking for a way to have remote autocomplete with vaadin-combo-box. The 'hot-autocomplete' is cool, but I don't think the AJAX mechanism should be bolted in the element. See e.g. how http://ellipticaljs.github.io/paper-autocomplete/ does it - just expose an API for changes and have the user of the component to provide the new suggestions.

**Comment by mercmobily** [comment]
I could do that, but it just looks a little too verbose for me. Did you
notice that in hot-autocomplete you can pass your own iron-ajax widget?
(Sorry vaadin people, I realise this is not the place to discuss this!)

**Comment by panuhorsmalahti** [comment]
Lets continue in https://github.com/mercmobily/hot-autocomplete/issues/1

**Comment by jouni** [comment]
Please add a size for this task since we want to track this work towards core elements 2.0

**Comment by manolo** [comment]
@panuhorsmalahti @mercmobily we are gathering info about what users expect about the components and what they use cases are. We might feel like the component is an 
enhanced replacement of the HTML select. But it maybe other ppl wanting to use it as a search box (like google search bar), highlight features etc.

Please could you give us some feedback? Thanks.

**Comment by panuhorsmalahti** [comment]
We're using vaadin-combo-box to display a text input into which the user can type custom strings or select a predefined value from the dropdown / autocomplete items list. In some cases we don't allow the user to type custom values at all.

We fetch the autocomplete items from the backend. We'd really like to dynamically fetch a smaller list of items depending on the user input, but we found out that this causes a few bugs. For now we're fetching all possible items (this is acceptable as so far the number of autocomplete items is fairly small).

**Comment by mercmobily** [comment]
Here is the list of features I feel are important:

* It should behave like a form element, exposing value
* Typing something should initiate the request to the server, and propose the first returned element as default (but "selected", so that more typing will neutralise it) AS WELL AS a list of possible options in a box
* Keyboard navigation for the box is crucial -- and difficult. I did it, but I feel it's quite brittle
* Cooperation with the server should happen via binding. I embeded an ajax request in my own widget to save lines of code, but it's not an ideal solution
* It needs to be possible to force the user to pick one of the proposed values (otherwise, the user should get a validation error), OR leave the autocompletion as "suggestion only"

I hope this helps!
